src/qibolab/_core/instruments/emulator/emulator.py

```diff
@@ -332,6 +363,8 @@ def extract_probabilities(
     Then, it computes probabilities, based on the identified
     expectations.
     """
+    expectations = [x.tidyup(atol=1e-2).diag() for x in expectations]
```

Comment:
Notice that the type hint is now wrong: `.tidyup()` is definitely not a method of a NumPy `NDArray`, but rather of `qutip.Qobj`.

I haven't yet taken the time to add type checking in the CI, just because I expect other issues from legacy code. But, if you can, just run some type checking in your IDE (eventually, it will be enforced).